apiserver/encrypting-data-at-datastore-layer.md

```diff
@@ -0,0 +1,334 @@
+
+#### State
+
+The distributed state machine is using the following data:
+
+1. keys named `key-<unisgned integer #>`, stored in `openshift-config-managed/<component>-encryption-<unsigned integer #>` secrets, and found via the label `encryption.apiserver.operator.openshift.io/component` for the respective component.
+2. the target encryption configuration stored in the `openshift-config-managed/encryption-config` secret as upstream [`apiserver.config.k8s.io/v1.EncryptionConfig`](https://github.com/kubernetes/kubernetes/blob/49891cc270019245a3d4796e84b33bf36d0bae08/staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go#L24) type (and synched to `<operand-target-namespace>/encryption-config`).
+3. `revision` label of running API server pods.
+4. the observed encryption configuration stored in the `<operand-target-namespace>/encryption-config-<revision>` secret.
+5. the encryption `APIServer` configuration defined above.
+
+Pod revisions and key numbers are unrelated.
+
+We say that a key `key-<n>` is
+
+1. **created** - if its secret `openshift-config-managed/<component>-encryption-<n>` is created.
+2. **configured read-key for resource GR** if it is defined as read-key for GR in the target encryption config secret. We call it **observed read-key for resource GR** if all API server instances are running with a corresponding config.
+3. **configured write-key for resource GR** if it is defined as write-key for GR in the target encryption config secret. We call it **observed write-key for resource GR** if all API server instances are running with a corresponding config.
+4. **migrated for resource GR** if the key secret `openshift-config-managed/<component>-encryption-<n>`'s annotation `encryption.operator.openshift.io/migrated-resources` lists the GR.
+5. **deleted** - if its secret `openshift-config-managed/<component>-encryption-<n>` is deleted.
+
+Each version of the operator has a fixed list of GRs that are supposed to be encrypted. All other GRs are **non-encrypted** for that operator. We say that such a resource of the former is **encrypted** if it is configured with at least a read-key in the target encryption config, and **to-be-encrypted** if it is not.
+
+#### Invariants
+
+The encryption configuration maintains the following invariants
+
+- the read-keys for all **encrypted** GRs are the same.
+- there is at most one non-identity write-key for all **encrypted** GRs (e.g. a new resource starts with `identity`, all others have a `aescbc` key already).
+
+#### Transitions
+
+The life-cycle of a key implemented by the controllers is: 1, 2, 3, 4, 5. 
+
+A transition only takes place when all API servers have converged to the same revision, are running, and hence use the same encryption configuration.
+
+- ->1: a new key secret is created if
+  
+  - encryption is being enabled via the API or
+  - a new **to-be-encrypted** resource shows up or
```

Comment:
huh, I don't remember creating a new key for this.

Reply:
> huh, I don't remember creating a new key for this.

oh, maybe I did.

Reply:
this was the core of your argument for the new one-key design: when another resource type must be encrypted, we create a new write key for **all** resources and push it through the state machine.